### PR TITLE
Http request/response DiagnosticSource logging.

### PR DESCRIPTION
**Pull request by nbilling**
- Added logging to WinHttpHandler, CurlHandler, and HttpHandlerToFilter
  for HTTP request/response.
- Added functional tests for HttpClient and WinHttpHandler.

@davidsh @cipop @stephentoub

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs [hunk 1]
```diff
@@ -3,9 +3,11 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Diagnostics;
 using System.Net;
 using System.Net.Http;
 using System.Net.Tests;
+using System.Reflection;
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs]: Reflection does NOT work in UWP apps against CoreFx libraries.  It is basically turned off since CoreFx UWP libraries (including System.Net.Http) build process apply the __BlockReflectionAttribute.

So, these tests which use Reflection will not work when we start having our GitHub tests execute against the netcore50 (UWP) versions of the libraries. Right now the CI system only executes dnxcore50 versions of the built library binaries.

So, I'm wondering is the use of Reflection here just a specific test thing or does the architecture of this logging depend on Reflection in general.  Because if it does depend architecture-wise on Reflection, I don't understand how any of this Diagnostic logging will work in the UWP scenarios.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs [hunk 2]
```diff
@@ -30,5 +32,115 @@ public void Timeout_SetTo60AndGetResponseFromServerWhichTakes40_Success()
...
+        [Fact]
+        public void SendAsync_ExpectedDiagnosticSourceNoLogging()
+        {
+            bool requestLogged = false;
+            bool responseLogged = false;
+
+            DiagnosticListener.AllListeners.Subscribe(new FakeDiagnosticListenerObserver(
+                kvp =>
+                {
+                    if (kvp.Key.Equals("System.Net.Http.Request"))
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs]: All of these strings such as "System.Net.Http.Request" should be factored out of these sources and into a Common file where we can track/rename all of them.  Please put them into a src/Common/src/System/Net/Http/* file and include it in the various CSPROJs.

**Comment by davidsh** [comment]
@vancem, can you please take a look at this.  I have concerns and questions about how Reflection is being used especially in the UWP scenarios where we block Reflection in our CoreFx libraries.

**Comment by nbilling** [comment]
@avanderhoorn

File: src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs (hunk 1 above)
Comment (nbilling) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs]: Is it possible to change the anonymous type so that it doesn't inherit the __BlockReflectionAttribute?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 3]
```diff
@@ -549,7 +552,9 @@ protected override void Dispose(bool disposing)
                 state,
                 CancellationToken.None,
                 TaskCreationOptions.DenyChildAttach,
-                TaskScheduler.Default);                
+                TaskScheduler.Default);
+
+            tcs.Task.ContinueWith(t => LogResponse(t, loggingRequestId));
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: This has allocations and should be done tracing is enabled. Further, it causes a closure that results allocations, so the fix needs take that into account.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 4]
```diff
@@ -83,6 +83,7 @@ public class WinHttpHandler : HttpMessageHandler
         private volatile bool _disposed;
         private SafeWinHttpHandle _sessionHandle;
         private WinHttpAuthHelper _authHelper = new WinHttpAuthHelper();
+        private static DiagnosticSource _diagnosticSource = new DiagnosticListener("HttpHandlerDiagnosticListener");
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: readonly

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 5]
```diff
@@ -1341,5 +1346,51 @@ private Task<bool> InternalReceiveResponseHeadersAsync(WinHttpRequestState state
 
             return state.TcsReceiveResponseHeaders.Task;
         }
+
+        private static Guid LogRequest(HttpRequestMessage request)
+        {
+            if (_diagnosticSource.IsEnabled("System.Net.Http.Request"))
+            {
+                Guid loggingRequestId = Guid.NewGuid();
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: I'd to see the body factored so that the simple check becomes inlineable.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 6]
```diff
@@ -1341,5 +1346,51 @@ private Task<bool> InternalReceiveResponseHeadersAsync(WinHttpRequestState state
 
             return state.TcsReceiveResponseHeaders.Task;
         }
+
+        private static Guid LogRequest(HttpRequestMessage request)
+        {
+            if (_diagnosticSource.IsEnabled("System.Net.Http.Request"))
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: These IDs should be in constants.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 7]
```diff
@@ -1341,5 +1346,51 @@ private Task<bool> InternalReceiveResponseHeadersAsync(WinHttpRequestState state
...
+            }
+
+            return Guid.Empty;
+        }
+
+        private static void LogResponse(Task<HttpResponseMessage> response, Guid loggingRequestId)
+        {
+            if (_diagnosticSource.IsEnabled("System.Net.Http.Response"))
+            {
+                if (!response.IsFaulted &&
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Ditto on all such logging methods.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs [hunk 8]
```diff
@@ -345,6 +349,9 @@ protected override void Dispose(bool disposing)
                 easy.FailRequest(exc);
                 easy.Cleanup(); // no active processing remains, so we can cleanup
             }
+
+            easy.Task.ContinueWith(t => LogResponse(t, loggingRequestId));
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: Same comments as earlier.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs [hunk 9]
```diff
@@ -640,6 +647,52 @@ private static void SetChunkedModeForSend(HttpRequestMessage request)
...
+                        new
+                        {
+                            Response = response.Result,
+                            LoggingRequestId = loggingRequestId,
+                            Timestamp = timestamp
+                        });
+                }
+            }
+        }
+
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: Can all of this code be factored out into a file/class/etc. that's just included wherever it's needed?

File: src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs [hunk 10]
```diff
@@ -265,5 +272,49 @@ private static HttpResponseMessage ConvertResponse(RTHttpResponseMessage rtRespo
...
+                      "System.Net.Http.Response",
+                      new
+                      {
+                          Response = response,
+                          LoggingRequestId = loggingRequestId,
+                          Timestamp = timestamp
+                      });
+                }
+            }
+        }
```
Comment (stephentoub) [inline src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs]: Does the logging really need to be done in the handler? Could it not be lifted out to HttpClient to share across all handlers?

File: src/System.Net.Http/src/project.json [hunk 11]
```diff
@@ -1,44 +1,47 @@
...
+                "System.Diagnostics.Tools": "4.0.0",
+                "System.Diagnostics.Tracing": "4.0.0",
+                "System.Net.Primitives": "4.0.10",
+                "System.Resources.ResourceManager": "4.0.0",
+                "System.Runtime.Extensions": "4.0.10",
+                "System.Runtime.WindowsRuntime": "4.0.10",
+                "System.Text.Encoding.Extensions": "4.0.10",
+                "System.Threading.Tasks": "4.0.10"
+            }
+        }
```
Comment (stephentoub) [inline src/System.Net.Http/src/project.json]: Looks like a bunch of formatting-only changes. Can you revert these?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 12]
```diff
@@ -1341,5 +1346,51 @@ private Task<bool> InternalReceiveResponseHeadersAsync(WinHttpRequestState state
...
+                    !response.IsCanceled &&
+                    // An empty loggingRequestId signifies that the request was not logged, so do
+                    // not attempt to log response.
+                    loggingRequestId != Guid.Empty)
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Why do we not log failed requests?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 13]
```diff
@@ -1341,5 +1346,51 @@ private Task<bool> InternalReceiveResponseHeadersAsync(WinHttpRequestState state
...
+            if (_diagnosticSource.IsEnabled("System.Net.Http.Request"))
+            {
+                Guid loggingRequestId = Guid.NewGuid();
+                long timestamp = Stopwatch.GetTimestamp();
+
+                _diagnosticSource.Write(
+                    "System.Net.Http.Request",
+                    new
+                    {
+                        Request = request,
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Is exposing  request and response objects like this really a good thing to do? It means any code in the process can hijack these objects, manipulate the streams, etc., right? That seems dangerous.

Comment (avanderhoorn) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: We have discussed this and that is side effect that we where prepared to accept. It is up to consumers to be responsible citizens and be responsible with what they done. There are logging use cases where mutations might occur (i.e. adding a header) but this isn't the average use case. @vancem

File: src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs (hunk 10 above)
Comment (avanderhoorn) [inline src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs]: :+1:

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 13 above)
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Thanks. Who is "we" here? This doesn't feel right to me. I'd get it at an app layer, but this is a lower level framework component, that will now be publishing private state for anyone to see and mess with. Am I just being paranoid?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Also, TaskScheduler.Default should be passed to ContinueWith as the scheduler argument.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 13 above)
Comment (vancem) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: This is one of those fundamental tradeoffs.   DiagnosticSource allows you to 'publish' lots of 'hooks' which show detailed state (note it should not be private class, but it is likely to be 'private' data).   Because we don't have a way of saying 'read only object' it also means that you CAN mutate things (although that is a abuse of DiagnosticSource to do so, the intention is that clients are effectively read-only).   

Thus DiagnosticSource creates a set of hooks that have 'scary power' like private reflection or the profiler API (both of which are not intended for program mutation).   It would be nice if it did not have this scary power to modify state, but there is no obvious way of doing that.    In a world of code access security Diagnostic source would have to be highly privileged.   

In short, we giving up on code access security (we trust all code in the process), and in that model this is OK.  

For what it is worth...

File: src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs (hunk 10 above)
Comment (nbilling) [inline src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs]: This is because we want to capture direct WinHttpHandler usage, on desktop, as well. If I log at both places then we end up logging the same event twice, and if I do something special to get around the double logging for desktop then we end up with a split between the platforms like I have here already.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs (hunk 9 above)
Comment (nbilling) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: That's a good question and something I wanted to ask the owners of this code re: adding a source file that would be shared between System.Net.Http and System.Net.WinHttpHandler builds.

**Comment by nbilling** [comment]
@delmyers

**Comment by delmyers** [comment]
@davidsh @vancem @avanderhoorn 
Can someone please clear up some of the confusion around what scenarios will be blocked by not having reflection available? Especially @avanderhoorn: It is my understanding that your current duck-typing mechanisms use reflection to generate code that implements interfaces to get at particular properties. Also, Is this a problem because of compiling to native for UWP? If the user doesn't compile to native does this problem go away (e.g. for debugging purposes)?

**Comment by vancem** [comment]
@delmyers Reflection is only an issue when compiling to Native.  UWP does not have anything to  do with it.    There is also the question of ReflectionEmit.   I believe Andrews duck typer uses Reflection emit, which is not support in .NET native, but I think he also has a way of calling back to using reflection only.   Andrew can speak to that.    Note that when we use CoreCLR for .NET Native debugging we do NOT have ReflectionEmit available (since it is trying to mimic the constraints of the .NET native tool chain)>

Now if we ignore the issue of ReflectionEmit (which we should be able  to do). the only issue with reflection is that the tool change LIKES TO OMMIT Metadata needed to reflect on things.   HOWEVER THIS IS UNDER OUR (Microsoft) CONTROL.   It is possible and even likely that the current rules as established might strip the meta-data for Networking DLL.   Having not performed the experiement, I can't say (I would bet it does).   It is likely that this could be fixed easily.

However it is UNCLEAR that always including the reflection meta-data for networking is a good tradeoff.   It is possible that tools like Glimpse can avoid refelction (simply cast the object to the object they expect).   This will work fine, it is just that they take a static dependency on that type.  For primitive types that are certain to be needed this is no big deal, so the problem is PROBLEM MUTE.  

However IF WE CARE ABOUT THE UWP Scenario (and it was unclear to me so far whether the answer to that is 'Yes')   AND the static casting solution is deemed inappropriate (again it seems like a good solution), then it seems pretty straigthfoward to simply pull in the experts on .NET Native reflection to craft a solution (which probably involves simply decorating something with the right custom attribute.

**Comment by davidsh** [comment]
In terms of the UWP scenario, it is not just NETNative that is affected.  It is also when NETNative is not used (i.e. F5 DEBUG experience building UWP apps).  Currently all System.* libraries (DLLs) for CoreFx use the __BlockReflectionAttribute when compiling the DLL for 'netcore50' target (i.e. UWP).  This is before any NETNative ILC compilation might be done on it.  This decision was made by the CoreFx team in order to reduce the memory/disk bloat of the libraries.

So, architecturely, if reflection is required on DLLs in order for DiagnosticSource to do its thing, then we have a problem.

And, to clarify things, there is more than one "Networking" dll.  There are several binaries involved such as System.Net.Sockets, System.Net.Http, System.Net.Http.WinHttpHandler, System.Net.Requests, etc.

**Comment by delmyers** [comment]
@vancem @davidsh Thanks for your responses. I seem to be getting conflicting details from each of you, though. @vancem seems to indicate that reflection is only a problem for .NET native, whereas @davidsh seems to indicate that it is a problem for all of UWP. If it is a problem for UWP in general, I would like it if we could come up with a solution that will not block potential glimpse and intellitrace UWP scenarios. @avanderhoorn , it is my understanding that you use reflection in order to discover properties and methods that you can emit types for use with duck typing. IntelliTrace may be able to get around the reflection issue by taking a dependency on a static type in our scenarios. From current email threads, though, I may be understanding that adding a static type is a little contentious.

**Comment by nbilling** [comment]
Regarding __BlockReflectionAttribute and UWP: Cristian pointed me towards the Runtime Directives config (https://msdn.microsoft.com/en-us/library/dn600639(v=vs.110).aspx) as a possible way to explicitly enable reflection for the types that we care about in System.Net.*. Does anyone know specifically how this interacts with __BlockReflectionAttribute and if we could use this to turn on metadata in a targeted way?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 5 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: I think I have done this now, but I am not really sure what the inlining heuristics of the C# compiler are. Let me know if it's still not inlineable and how I should fix it.

**Comment by nbilling** [comment]
@davidsh @delmyers @vancem @avanderhoorn @stephentoub 
Ping. I'm happy with the resolution of the issues raised so far and from the IntelliTrace side we are good to merge now. Regarding the reflection in a functional test thing: if/when the tests starts failing in UWP builds we would like to take that to the project N team as a bug on them, as conceptually it makes sense to emit metadata for anonymous types by default.

If there are other objections please raise them now :)

File: src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                        LoggingRequestId = loggingRequestId,
+                        Timestamp = timestamp
+                    }
+                );
+
+                return loggingRequestId;
+            }
+
+            return Guid.Empty;
+        }
```
Comment (stephentoub) [inline src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs]: Regarding my comment about inlining, from a perf perspective it may of course not matter, but my intention was that this be changed to something like:
```C#
public static Guid LogHttpRequest(this DiagnosticListener @this, HttpRequestMessage request)
{
    return @this.IsEnabled(HttpHandlerLoggingStrings.RequestWriteName) ?
        LogHttpRequestCore(@this, request) :
        Guid.Empty;
}
```
and then having all of the work in a private method that's delegated to.  I'm fine with you going ahead without that, as given everything else that's going on, it's unlikely to make a difference.  We can measure it after (for reference, you can use PerfView to see whether inlining is happening or not; it has a JIT Inlining feature that'll show you everything that's inlined, everything that's not, and why).

File: src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            responseTask.ContinueWith(
+                t =>
+                {
+                    if (!t.IsFaulted &&
+                        !t.IsCanceled)
+                    {
+                        LogHttpResponseInner(diagnosticListener, t.Result, loggingRequestId);
+                    }
+                },
+                TaskScheduler.Default);
```
Comment (stephentoub) [inline src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs]: How important is performance for the situation where logging is enabled?  I'm assuming "not very", in which case this is fine.  If that's not the answer, there are some tweaks that can be made here to avoid allocations and the like.

File: src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/System.Net.Http.WinHttpHandler.Functional.Tests.csproj [hunk 16]
```diff
@@ -31,5 +31,8 @@
   <ItemGroup>
     <None Include="project.json" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Include="FakeDiagnosticListenerObserver.cs" />
+  </ItemGroup>
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/System.Net.Http.WinHttpHandler.Functional.Tests.csproj]: Why is this in its own ItemGroup down here?

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs [hunk 17]
```diff
@@ -46,6 +46,8 @@ internal partial class CurlHandler : HttpMessageHandler
         private readonly static bool s_supportsSSL;
         private readonly static bool s_supportsHttp2Multiplexing;
 
+        private readonly static DiagnosticListener _diagnosticListener = new DiagnosticListener(HttpHandlerLoggingStrings.DiagnosticListenerName);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: Nit: statics should have an ```s_``` prefix

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs [hunk 18]
```diff
@@ -83,6 +83,7 @@ public class WinHttpHandler : HttpMessageHandler
         private volatile bool _disposed;
         private SafeWinHttpHandle _sessionHandle;
         private WinHttpAuthHelper _authHelper = new WinHttpAuthHelper();
+        private static readonly DiagnosticListener _diagnosticListener = new DiagnosticListener(HttpHandlerLoggingStrings.DiagnosticListenerName);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Nit: statics should have an ```s_``` prefix

File: src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs [hunk 19]
```diff
@@ -25,6 +25,8 @@ namespace System.Net.Http
 {
     internal class HttpHandlerToFilter : HttpMessageHandler
     {
+        private readonly static DiagnosticListener _diagnosticListener = new DiagnosticListener(HttpHandlerLoggingStrings.DiagnosticListenerName);
```
Comment (stephentoub) [inline src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs]: Nit: statics should have an ```s_``` prefix

File: src/System.Net.Http.WinHttpHandler/src/project.json [hunk 20]
```diff
@@ -1,6 +1,7 @@
 {
   "dependencies": {
     "Microsoft.Win32.Primitives": "4.0.0",
+    "System.Diagnostics.DiagnosticSource": "4.0.0-rc3-23729",
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/project.json]: I'm surprised this passes CI... I'd expect these 23729 versions numbers to cause failures due to the package verification.  I believe we're currently on 23808.

File: src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs [hunk 21]
```diff
@@ -0,0 +1,71 @@
+﻿using System.Collections.Generic;
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs]: Why do we need this type in source multiple times? In general for these tests, I'd expect it'd be sufficient to just have them for System.Net.Http, as that'll end up covering both the WinHttpHandler and CurlHandler when running those functional tests on each of the platforms.  If we do for some reason need to have separate duplicate tests for WinHttpHandler, we should share the source rather than duplicating it.

File: src/System.Net.Http.WinHttpHandler/src/project.json (hunk 20 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/project.json]: Should I update these to 23808? I'm not sure of what the strategy is with checking in specific build dependencies. There is no way to take a dependency on "the version that will be built alongside this project" is there?

File: src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs (hunk 21 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs]: That's a good point. I will remove the winhttphandler functional tests since, as you say, that exact same code path is exercised by the httclient functional tests.

File: src/System.Net.Http.WinHttpHandler/src/project.json (hunk 20 above)
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/project.json]: Yes.  We update them en mass to move all values in the repo to the next known-good set of packages.
https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/project-nuget-dependencies.md#upgrading-a-package-dependency

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by nbilling** [comment]
Eh, I didn't actually mean to push that most recent commit on its own. If anyone is reading stuff after this comment then maybe hold off until I make another small tweak.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs (hunk 1 above)
Comment (nbilling) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs]: As mentioned in another comment-- our preference is to continue with the expectation that project N will emit metadata for anonymous types and then file a bug against them, for instance, if the HttpClient functional tests started failing for UWP.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs (hunk 2 above)
Comment (nbilling) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 3 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 4 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 6 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 7 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Fixed.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs (hunk 8 above)
Comment (nbilling) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: Fixed.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs (hunk 9 above)
Comment (nbilling) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: Fixed (shared code for this moved to :/src/common/src/system/net/http/httphandlerdiagnosticlistenerextension.cs).

File: src/System.Net.Http/src/project.json (hunk 11 above)
Comment (nbilling) [inline src/System.Net.Http/src/project.json]: Fixed.

File: src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs (hunk 14 above)
Comment (nbilling) [inline src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs]: Fixed.

File: src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs (hunk 15 above)
Comment (nbilling) [inline src/Common/src/System/Net/Http/HttpHandlerDiagnosticListenerExtensions.cs]: My impression is that the answer is closer to "not very" than the alternative, but at the same time I don't have anything like a spec that says "perf isn't important."

File: src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/System.Net.Http.WinHttpHandler.Functional.Tests.csproj (hunk 16 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/System.Net.Http.WinHttpHandler.Functional.Tests.csproj]: Fixed (in the sense that all changes under :/src/system.net.http.winhttphandler/tests were reverted).

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs (hunk 17 above)
Comment (nbilling) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs (hunk 18 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Fixed.

File: src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs (hunk 19 above)
Comment (nbilling) [inline src/System.Net.Http/src/netcore50/System/Net/HttpHandlerToFilter.cs]: Fixed.

File: src/System.Net.Http.WinHttpHandler/src/project.json (hunk 20 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/src/project.json]: Fixed.

File: src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs (hunk 21 above)
Comment (nbilling) [inline src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs]: Fixed.

**Comment by nbilling** [comment]
Everything should be addressed and good to go.

File: src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs (hunk 21 above)
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/FakeDiagnosticListenerObserver.cs]: The file is still in the PR.

**Comment by stephentoub** [comment]
One remaining issue (a file that shouldn't be there), then LGTM.

Can you please squash this and rebase off master?  There are conflicts to be resolved.

Thanks.

**Comment by nbilling** [comment]
Okay rebased, removed file, and squashed all the PR feedback commits :)

**Comment by nbilling** [comment]
It seems like changing those version numbers has caused some kind of mayhem. Investigating...

**Comment by nbilling** [comment]
Actually the issue is a change that was made that refactored the system.net.http project.json file into 3 platform specific files, rebasing off dotnet/master brought in the commit. I think I have it fixed but rebuilding to make sure it's all good.

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by davidsh** [comment]
@stephentoub This PR seems stuck and the CI system won't start on it.  Does the PR need to be closed and re-opened? Or is something else wrong with CI?

**Comment by stephentoub** [comment]
Jenkins interaction with the repo is broken again.  All of the recent PRs aren't having builds done.  @mmitche is looking into it.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by davidsh** [comment]
CI Failure in Windows tests:

```
Assert.Throws() Failure\r\nExpected: typeof(System.Net.Http.HttpRequestException)\r\nActual: typeof(System.IO.FileNotFoundException): Could not load file or assembly 'System.Diagnostics.DiagnosticSource, Version=4.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
```
